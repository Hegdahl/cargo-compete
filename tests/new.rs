pub mod common;

use ignore::overrides::Override;
use insta::{assert_json_snapshot, assert_snapshot};
use liquid::object;
use std::io::BufRead;

#[test]
fn atcoder_agc047() -> anyhow::Result<()> {
    let (output, tree) = run(&b""[..], "agc047")?;
    assert_snapshot!("atcoder_agc047_output", output);
    assert_json_snapshot!("atcoder_agc047_file_tree", tree, { r#".**["Cargo.lock"]"# => ".." });
    Ok(())
}

#[cfg(feature = "__test_with_credentials")]
#[test]
fn atcoder_practice() -> anyhow::Result<()> {
    let (output, tree) = run(common::atcoder_credentials()?, "practice")?;
    assert_snapshot!("atcoder_practice_output", output);
    assert_json_snapshot!("atcoder_practice_file_tree", tree, { r#".**["Cargo.lock"]"# => ".." });
    Ok(())
}

fn run(
    input: impl BufRead + 'static,
    contest: &str,
) -> anyhow::Result<(String, serde_json::Value)> {
    common::run(
        |workspace_root| -> _ {
            std::fs::create_dir_all(workspace_root.join("cargo-compete-template").join("src"))?;

            std::fs::write(
                workspace_root.join("Cargo.toml"),
                r#"[workspace]
members = ["cargo-compete-template"]
"#,
            )?;

            std::fs::write(
                workspace_root.join("compete.toml"),
                liquid::ParserBuilder::with_stdlib()
                    .build()?
                    .parse(include_str!("../resources/compete.toml.liquid"))?
                    .render(&object!({
                        "template_platform": "atcoder",
                        "submit_via_binary": false,
                    }))?,
            )?;

            std::fs::write(
                workspace_root
                    .join("cargo-compete-template")
                    .join("Cargo.toml"),
                r#"[package]
name = "cargo-compete-template"
version = "0.1.0"
edition = "2018"
"#,
            )?;

            std::fs::write(
                workspace_root
                    .join("cargo-compete-template")
                    .join("src")
                    .join("main.rs"),
                r#"fn main() {
    todo!();
}
"#,
            )?;
            Ok(())
        },
        input,
        &["", "compete", "n", contest],
        |workspace_root, output| {
            output
                .replace(workspace_root.to_str().unwrap(), "{{ cwd }}")
                .replace('/', "{{ slash_or_backslash }}")
                .replace('\\', "{{ slash_or_backslash }}")
        },
        |_| Ok(Override::empty()),
    )
}
