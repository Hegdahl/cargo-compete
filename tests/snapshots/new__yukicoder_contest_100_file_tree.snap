---
source: tests/new.rs
expression: tree
---
{
  ".cargo": {
    "config.toml": "[cargo-new]\nname = \"\"\nemail = \"\"\n"
  },
  "compete.toml": "test-suite = \"{{ manifest_dir }}/testcases/{{ bin_alias | kebabcase }}.yml\"\n\n[new]\nplatform = \"yukicoder\"\npath = \"./{{ package_name }}\"\n\n[new.template.dependencies]\nkind = \"inline\"\ncontent = '''\nproconio = \"=0.3.6\"\n'''\n\n[new.template.src]\nkind = \"inline\"\ncontent = '''\nfn main() {\n    todo!();\n}\n'''\n",
  "contest100": {
    "Cargo.toml": "[package]\nname = \"contest100\"\nversion = \"0.1.0\"\nauthors = [\"\"]\nedition = \"2018\"\n\n[package.metadata.cargo-compete.bin]\ncontest100-a = { alias = \"a\", problem = \"https://yukicoder.me/problems/no/191\" }\ncontest100-b = { alias = \"b\", problem = \"https://yukicoder.me/problems/no/192\" }\ncontest100-c = { alias = \"c\", problem = \"https://yukicoder.me/problems/no/193\" }\ncontest100-d = { alias = \"d\", problem = \"https://yukicoder.me/problems/no/194\" }\ncontest100-e = { alias = \"e\", problem = \"https://yukicoder.me/problems/no/195\" }\ncontest100-f = { alias = \"f\", problem = \"https://yukicoder.me/problems/no/196\" }\n\n[[bin]]\nname = \"contest100-a\"\npath = \"src/bin/a.rs\"\n\n[[bin]]\nname = \"contest100-b\"\npath = \"src/bin/b.rs\"\n\n[[bin]]\nname = \"contest100-c\"\npath = \"src/bin/c.rs\"\n\n[[bin]]\nname = \"contest100-d\"\npath = \"src/bin/d.rs\"\n\n[[bin]]\nname = \"contest100-e\"\npath = \"src/bin/e.rs\"\n\n[[bin]]\nname = \"contest100-f\"\npath = \"src/bin/f.rs\"\n[dependencies]\nproconio = \"=0.3.6\"\n",
    "src": {
      "bin": {
        "a.rs": "fn main() {\n    todo!();\n}\n",
        "b.rs": "fn main() {\n    todo!();\n}\n",
        "c.rs": "fn main() {\n    todo!();\n}\n",
        "d.rs": "fn main() {\n    todo!();\n}\n",
        "e.rs": "fn main() {\n    todo!();\n}\n",
        "f.rs": "fn main() {\n    todo!();\n}\n"
      }
    },
    "testcases": {
      "a.yml": "---\ntype: Batch\ntimelimit: 5s\nmatch: Lines\n\ncases:\n  - name: sample1\n    in: |\n      4\n      500 300 200 100\n    out: |\n      30\n  - name: sample2\n    in: |\n      5\n      400 250 200 100 50\n    out: |\n      60\n  - name: sample3\n    in: |\n      5\n      200 200 200 200 200\n    out: |\n      0\n\nextend: []\n",
      "b.yml": "---\ntype: Batch\ntimelimit: 2s\nmatch: Lines\ncases:\n  - name: sample1\n    in: \"101\\n\"\n    out: ~\n    timelimit: ~\n    match: ~\n  - name: sample2\n    in: \"1000\\n\"\n    out: ~\n    timelimit: ~\n    match: ~\nextend: []",
      "c.yml": "---\ntype: Batch\ntimelimit: 1s\nmatch: Lines\n\ncases:\n  - name: sample1\n    in: |\n      1+1\n    out: |\n      2\n  - name: sample2\n    in: |\n      1+111\n    out: |\n      112\n  - name: sample3\n    in: |\n      10+4-1+15\n    out: |\n      514\n\nextend: []\n",
      "d.yml": "---\ntype: Batch\ntimelimit: 5s\nmatch: Lines\n\ncases:\n  - name: sample1\n    in: \"2 5\\n1 1\"\n    out: 5 12\n  - name: sample2\n    in: \"5 10\\n1 2 3 4 5\"\n    out: 214 438\n  - name: sample3\n    in: \"30 987654321012\\n3 1 4 1 5 9 2 6 5 3 5 8 9 7 9 3 2 3 8 4 6 2 6 4 3 3 8 3 2 7\"\n    out: 923032656 920866414\n\nextend: []\n",
      "e.yml": "---\ntype: Batch\ntimelimit: 5s\nmatch: Lines\n\ncases:\n  - name: sample1\n    in: 5 13 34\n    out: 1 1\n  - name: sample2\n    in: 11 11 11\n    out: 1 3\n  - name: sample3\n    in: 5 6 9\n    out: \"-1\"\n  - name: sample4\n    in: 35169 76228629 17995137\n    out: 5487 14841\n\nextend: []\n",
      "f.yml": "---\ntype: Batch\ntimelimit: 2s\nmatch: Lines\n\ncases:\n  - name: sample1\n    in: |\n      6 1\n      0 1\n      0 2\n      1 3\n      2 4\n      2 5\n    out: |\n      3\n  - name: sample2\n    in: |\n      6 2\n      0 1\n      0 2\n      1 3\n      2 4\n      2 5\n    out: |\n      4\n  - name: sample3\n    in: |\n      6 3\n      0 1\n      0 2\n      1 3\n      2 4\n      2 5\n    out: |\n      4\n\nextend: []\n"
    }
  }
}
